src/python/pants/backend/jvm/tasks/jvm_compile/rsc/rsc_compile.py

```diff
@@ -390,6 +413,102 @@ def desandboxify_pantsd_loc(path):
+      # picks the cached distribution
+      distribution = JvmPlatform.preferred_jvm_distribution([], strict=True)
+      jvm_lib_jars_abs = distribution.find_libs(['rt.jar', 'dt.jar', 'jce.jar', 'tools.jar'])
+      cp_entries.extend(jvm_lib_jars_abs)
+
+      # TODO use compile_classpath
+      classpath_abs = [
+        path for (conf, path) in
+        self.context.products.get_data('rsc_classpath').get_for_target(ctx.target)
+      ]
+      dependency_classpath = self._zinc.compile_classpath(
+        'compile_classpath',
+        ctx.target,
+        extra_cp_entries=self._extra_compile_time_classpath)
+      classpath_rel = fast_relpath_collection(classpath_abs)
+      cp_entries.extend(classpath_rel)
+
+      counter_val = str(counter()).rjust(counter.format_length(), b' ')
+      counter_str = '[{}/{}] '.format(counter_val, counter.size)
+      # Do some reporting.
+      self.context.log.info(
+        counter_str,
+        'Metacp-ing ',
+        items_to_report_element(cp_entries, 'jar'),
+        ' in ',
+        items_to_report_element([t.address.reference() for t in vts.targets], 'target'),
+        ' (',
+        ctx.target.address.spec,
+        ').')
+
+      ctx.ensure_output_dirs_exist()
+
+      tgt, = vts.targets
+      with Timer() as timer:
+      # Step 1: Convert classpath to SemanticDB
+        # ---------------------------------------
+        scalac_classpath_path_entries_abs = self.tool_classpath('workaround-metacp-dependency-classpath')
+        scalac_classpath_path_entries = fast_relpath_collection(scalac_classpath_path_entries_abs)
+        rsc_index_dir = fast_relpath(ctx.rsc_index_dir, get_buildroot())
+        args = [
+          '--verbose',
+          # NB: Without this setting, rsc will be missing some symbols
+          #     from the scala library.
+          '--include-scala-library-synthetics', # TODO generate these once and cache them
+          # NB: We need to add these extra dependencies in order to be able
+          #     to find symbols used by the scalac jars.
+          '--dependency-classpath', os.pathsep.join(dependency_classpath + scalac_classpath_path_entries),
+          # NB: The directory to dump the semanticdb jars generated by metacp.
+          # TODO: break this out into a separate job so that we can apply
+          #       it once per 3rdparty target.
+          #       https://github.com/pantsbuild/pants/issues/6515
+          '--out', rsc_index_dir,
+          os.pathsep.join(cp_entries),
+        ]
+        metacp_wu = self._runtool(
+          'scala.meta.cli.Metacp',
+          'metacp',
+          args,
+          distribution,
+          tgt=tgt,
+          input_files=(scalac_classpath_path_entries + classpath_rel),
+          output_dir=rsc_index_dir)
+        metacp_stdout = stdout_contents(metacp_wu)
+        metacp_result = json.loads(metacp_stdout)
+
+        metai_classpath = self._collect_metai_classpath(
+          metacp_result, classpath_rel, jvm_lib_jars_abs)
+
```

Comment:
Before we land this, I think we should incorporate at least one `@rule` to begin weaning ourselves off of the use of the execution graph.

Reply:
I'd prefer to have introducing `@rule` be a separate PR from this. I could bail on this change and start an `@rule` conversion before landing it, or I could fix this up and land it and then introduce a migration towards using `@rule`s.

Reply:
@baroquebobcat : I agree that getting this green and landable and then working on `@rule`s in a separate commit would be a good idea. Depending on how that goes, you can either land them as separate PRs, or land it here.